src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -1794,73 +1848,192 @@ int TheoryStringsRewriter::componentContains(std::vector<Node>& n1,
     unsigned diff = n1.size() - n2.size();
     for (unsigned i = 0; i <= diff; i++)
     {
-      bool check = false;
-      if (n1[i] == n2[0])
-      {
-        check = true;
-      }
-      else if (n1[i].isConst() && n2[0].isConst())
-      {
-        // could be suffix
-        CVC4::String s = n1[i].getConst<String>();
-        CVC4::String t = n2[0].getConst<String>();
-        if (t.size() < s.size() && s.suffix(t.size()) == t)
-        {
-          check = true;
-        }
-      }
-      if (check)
+      Node n1rb_first;
+      Node n1re_first;
+      // first component of n2 must be a suffix
+      if (componentContainsBase(n1[i],
+                                n2[0],
+                                n1rb_first,
+                                n1re_first,
+                                1,
+                                computeRemainder && remainderDir != 1))
       {
-        bool success = true;
+        Assert(n1re_first.isNull());
         for (unsigned j = 1; j < n2.size(); j++)
         {
-          if (n1[i + j] != n2[j])
+          // are we in the last component?
+          if (j + 1 == n2.size())
           {
-            if (j + 1 == n2.size() && n1[i + j].isConst() && n2[j].isConst())
+            Node n1rb_last;
+            Node n1re_last;
+            // last component of n2 must be a prefix
+            if (componentContainsBase(n1[i + j],
+                                      n2[j],
+                                      n1rb_last,
+                                      n1re_last,
+                                      -1,
+                                      computeRemainder && remainderDir != -1))
             {
-              // could be prefix
-              CVC4::String s = n1[i + j].getConst<String>();
-              CVC4::String t = n2[j].getConst<String>();
-              if (t.size() < s.size() && s.prefix(t.size()) == t)
+              Assert(n1rb_first.isNull());
```

Comment:
Is this really true? What if you have something like:
```
n1 = "abc", "def"
n2 = "bc", "de"
```
And you call `componentContains()` with `remainderDir = -1`. Then you should get `n1rb_first = "a"`, no?

Reply:
Ah yes this meant to be n1rb_last.isNull()